src/OrleansAzureUtils/Storage/AzureTableDataManager.cs

```diff
@@ -123,22 +123,27 @@ public AzureTableDataManager(string tableName, string storageConnectionString, L
         /// <returns>Completion promise for this operation.</returns>
         public async Task ClearTableAsync()
         {
-            IEnumerable<Tuple<T,string>> items = (await ReadAllTableEntriesAsync()).ToList();
-            List <Tuple<T,string>> batch = new List<Tuple<T, string>>(AzureTableDefaultPolicies.MAX_BULK_UPDATE_ROWS);
-            foreach (Tuple<T, string> item in items)
+            IEnumerable<Tuple<T,string>> items = await ReadAllTableEntriesAsync();
+            IEnumerable<IGrouping<string, Tuple<T, string>>> partitions = items.GroupBy(item => item.Item1.PartitionKey);
+            List<Tuple<T,string>> batch = new List<Tuple<T, string>>(AzureTableDefaultPolicies.MAX_BULK_UPDATE_ROWS);
+            foreach (IGrouping<string, Tuple<T, string>> partition in partitions)
             {
-                batch.Add(item);
-                // delete and clear when we have a full batch
-                if (batch.Count == AzureTableDefaultPolicies.MAX_BULK_UPDATE_ROWS)
+                foreach (Tuple<T, string> item in partition)
+                {
+                    batch.Add(item);
+                    // delete and clear when we have a full batch
+                    if (batch.Count == AzureTableDefaultPolicies.MAX_BULK_UPDATE_ROWS)
+                    {
+                        await DeleteTableEntriesAsync(batch);
+                        batch.Clear();
+                    }
+                }
+                if (batch.Count != 0)
```

Comment:
Is it worth parallelizing? Also, it might be work creating a small extension method for batching, to make it more readable (and also easier to parallelize if that's appropriate.
See http://stackoverflow.com/questions/13731796/create-batches-in-linq for very simple extensions